ignite/metrics/mean_average_precision.py

```diff
@@ -0,0 +1,664 @@
+                metric's device to be the same as your ``update`` arguments ensures the ``update`` method is
+                non-blocking. By default, CPU.
+        """
+        if rec_thresholds is not None:
+            self.rec_thresholds: Optional[torch.Tensor] = self._setup_thresholds(rec_thresholds, "rec_thresholds")
+        else:
+            self.rec_thresholds = None
+
+        if average_operand not in ("precision", "max-precision"):
+            raise ValueError(f"Wrong `average_operand` parameter, given {average_operand}")
+        self.average_operand = average_operand
+
+        if class_mean is not None and class_mean not in ("micro", "macro", "weighted", "with_other_dims"):
+            raise ValueError(f"Wrong `class_mean` parameter, given {class_mean}")
+        self.class_mean = class_mean
+
+        self.allow_multiple_recalls_at_single_threshold = allow_multiple_recalls_at_single_threshold
+
+        super(_BasePrecisionRecall, self).__init__(
+            output_transform=output_transform, is_multilabel=classification_is_multilabel, device=device
+        )
+
+        if self._task == "classification" and self.class_mean == "with_other_dims":
+            raise ValueError("class_mean 'with_other_dims' is not compatible with classification.")
+
+    def _setup_thresholds(self, thresholds: Union[Sequence[float], torch.Tensor], threshold_type: str) -> torch.Tensor:
+        if isinstance(thresholds, Sequence):
+            thresholds = torch.tensor(thresholds)
+
+        if isinstance(thresholds, torch.Tensor):
+            if thresholds.ndim != 1:
+                raise ValueError(
+                    f"{threshold_type} should be a one-dimensional tensor or a sequence of floats"
+                    f", given a {thresholds.ndim}-dimensional tensor."
+                )
+            thresholds = thresholds.sort().values
+        else:
+            raise TypeError(f"{threshold_type} should be a sequence of floats or a tensor, given {type(thresholds)}.")
+
+        if min(thresholds) < 0 or max(thresholds) > 1:
+            raise ValueError(f"{threshold_type} values should be between 0 and 1, given {thresholds}")
+
+        return cast(torch.Tensor, thresholds)
+
+    @reinit__is_reduced
+    def reset(self) -> None:
+        """
+        Reset method of the metric
+        """
+        super(_BasePrecisionRecall, self).reset()
+        if self.do_matching.__func__ == MeanAveragePrecision.do_matching:  # type: ignore[attr-defined]
+            self._task: Literal["classification", "detection"] = "classification"
+        else:
+            self._task = "detection"
+        self._tp = defaultdict(lambda: [])
+        self._fp = defaultdict(lambda: [])
+        if self._task == "detection":
+            self._scores = defaultdict(lambda: [])
+            self._P = defaultdict(lambda: 0)
+            self._num_classes = 0
+        else:
+            self._scores = []
+            self._P = []
```

Comment:
Can we have the same code for both tasks to avoid these if/else clauses. 
If we can't unify the structures for some reasons, let's override `reset` method in ObjectDetectionMAP.